### DISCUSSION TITLE
Error 429

### DISCUSSION THREAD

**Discussion by bostonsam** [tweepy/tweepy#1820] [Questions]
Hey, guys, I have few questions for you

I'm trying to build a bot for twitter with API keys and tweepy 
I'm getting an error 429 and for what I know for this, I must refer to twitter rate limits, so I go on the page : https://developer.twitter.com/en/docs/basics/rate-limits
And as I can read my  rate limit should be 1000 tweet liked a for 24h 
Good so I start coding and this is my code

`api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

user = api.me()

search = 'best food'
nrTweets = 1000

for tweet in tweepy.Cursor(api.search, search).items(nrTweets):
    try:

        print('Tweet Liked')
        tweet.favorite()

    except tweepy.TweepError as e:
        print(e.reason)
        print("paused for 15min")
        time.sleep(15*60)`

Now this way I tried to avoid the 429 problems by waiting 15 min 
But my question are 2

Why Am I getting this error in the first place, I mean I should be able to pull 1000 likes instead I stop at few? 
Second, how can I target only the error 429 
With my code, and is no surprise to me, I get a pause for all the errors so even if I encounter a tweet already liked I get to wait 15 min 

I thought that I could use an if loop but because I'm a newbie I don't know how to get the syntax right. Can you give me a specific example of that, please?

As I said I'm a complete noob as a matter of fact I'm coding from 3 days and most of my code although I understand most of it is copied so sorry if this is a stupid question, but I'm not able to find my solution by myself yet so I thought to ask you guys.

My goal for the bot is to get as many tweets as I can liked without being banned

**Comment by MichaelCurrin** [comment]
The status code 429 means too many requests and happens for rate limit exceeding, as you have gathered.

You are correct in saying 1000 likes allowed for 24 hours and your limit of 1000 tweets takes this into account.

You don't need to add the sleep part. In the case of liking you'd have to wait 24 hours - so sleeping 15 minutes won't help.

Your problem might be liking a tweet that has already been liked. So I'd say check `tweet.favorited is False` before liking something.

Regarding targeting a 429 error, you can use `tweepy.RateLimitError` on place of `tweepy.TweepError`. Or catch them separately. That alone won't help though as the rate limit error will be caught by TweepError too. 

Try reduce your limit to say 100 and see if it okay. And then scale it up

**Comment by bostonsam** [comment]
I actually get the error but the code skyp that like and continue,
I will try Tweepy.RateLimitError can I do something like this ?

`search = 'best food'
nrTweets = 100


for tweet in tweepy.Cursor(api.search, search).items(nrTweets):
    try:

        if tweet.favorited == False:
            tweet.favorite()
            print('Tweet Liked')

    except tweepy.RateLimitError as e:
        print(e.reason)`

**Comment by MichaelCurrin** [comment]
Oh I thought TweepError was catching it so then yes update your code as you suggested.

BTW use triple backtick for code blocks not single and also keep the backticks on their own line. Optionally specify language like python for highlighting 


    ```python
    print("Hello")
    ```

Renders as

```python
print("Hello")
```

**Comment by bostonsam** [comment]
Thanks actually Ratelimiterror doesn't work for me it stops the interaction if i get an already favourited tweet

Now I modified my code like this 

```python

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

user = api.me()

search = 'Italian dish'
nrTweets = 200


for tweet in tweepy.Cursor(api.search, search).items(nrTweets):

    try:
        if tweet.favorited is False:
            tweet.favorite()
            print('Tweet Liked')

    except tweepy.TweepError as e:

        print(e.reason)
    except StopIteration:
        break

```
This gives me an ok result but as soon as I try to rerun the code I get 429 error, it seems to me that there is something else limiting my possibility of favouring
after only 200 likes I get the error or even less I don't know for sure (I'm hypothesizing because at a certain point the code stop and I think that it means it reaches the query nrTweets), also I see that for some 429 errors if I wait 15min let me restart the liking and that to me, that means that I have other restriction than the "tweet favourite" one.
Finally it doesn't stop for 429 error of course with my code but only print the reason.

Also the statement "if tweet.favorited is False:" don't do much for me, i get the same result with or without it 

My question is this, with the code you see now am I making some other request that could have Rate limiting?

**Comment by bostonsam** [comment]
Maybe I found why can You tell me if is this what I'm looking for, i think so
https://developer.twitter.com/en/docs/basics/rate-limiting

**Comment by MichaelCurrin** [comment]
Sorry I don't know.

Regarding rate limits, Twitter will give you a listing of _all_ your limits by endpointing, covering how many requests are left and when they reset. I haven't used it in a while but the API docs show output example.

https://developer.twitter.com/en/docs/developer-utilities/rate-limit-status/api-reference/get-application-rate_limit_status

How to do it in Tweepy, based on [StackOverflow answer](stackoverflow.com/questions/30313124/python-2-7-tweepy-how-to-get-rate-limit-status).

```python
data = api.rate_limit_status()

# Print everything
print data['resources']

# Print search limits
print data['resources']['statuses']['/statuses/search']
```
I can't find how to do favorites, so I added the printing of everything.

You don't even have to do it as part of your script. You can run this bit above in a standalone script (with auth setup) - run it every few minutes or just after you've hit a fatal error.

That requires an extra request of course. Another way to do it is to get the rate limit status on the response object as part of the search, like, etc. I forget how to do that.

**Comment by Harmon758** [comment]
It's unclear what's going on here. You'll have to provide a full traceback and explain your expected and actual behavior if you want further help.

You can catch [`tweepy.TooManyRequests`](https://tweepy.readthedocs.io/en/v4.5.0/exceptions.html#tweepy.errors.TooManyRequests) to handle 429 HTTP errors specifically.

An "if loop" is not a thing. An `if` statement uses a conditional and is not a loop.

`tweet.favorited` will always be `False`, as the endpoint that [`API.search_tweets`](https://tweepy.readthedocs.io/en/v4.5.0/api.html#tweepy.API.search_tweets) (renamed from `API.search` in Tweepy v4.0.0) uses does not support perspectival attributes (fields that pertain to the perspective of the authenticating user). Instead, you can handle the [`tweepy.Forbidden`](https://tweepy.readthedocs.io/en/v4.5.0/exceptions.html#tweepy.errors.Forbidden) exception that's raised with the 139 API error code.

For code block usage, see https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks.

**Comment by SodiKroehler** [comment]
I have a similar error, and found a workaround on the twitter labs. Sharing it here in hopes that perhaps it could be added as a feature?
v2 API throws a 429 error if the rate limit is reached per fifteen minutes or if the rate limit is reached per second. Tweepy v2 `Client` has an option to `wait_on_rate_limit`, which if set, provides functionality for catching the 15 minute rate limit by simply sleeping for the `x-rate-limit-reset` in the header. In the case of the second cause of the 429 error, that will pretty much always be the maximum 15 minutes, allowing you to get maybe one or two requests per 15 minutes. Workaround that I did is as follows:
``` 
try:
    my_client.search_all_tweets
    retryCount = 0
    ...
except tweepy.errors.TooManyRequests:
    if (retryCount < 10):
        sleep(1)
        continue
    else:
        sleep(900)
        print("429 with too many retries, sleeping for 15 mins")
        continue
```
The `resetCount` is reset after every successful call, but still prevents rate limit abuse by stopping after 10 tries. Another possible workaround would be just sleeping for 1 second after every request. Either way, it seems like the `wait_on_rate_limit` option in the client should accommodate this.
